### PR TITLE
fix #4106: proprietary oauth

### LINKED ISSUE #4106: Authorize codes via OAuth with expiry
Allow code roles to be authorized via OAuth. Sirepo will ask the OAuth server to validate user credentials. The OAuth server will return an access token with an expiry. 

The OAuth handoff will need to describe where it is going so that needs to be configurable on a per code basis.

Roles need an expiry and a description of an authentication method. I don't now that we can reuse sirepo.auth for this, but this is so specific that I think we can just refactor sirepo.auth.github module to support this workflow.

### PR DESCRIPTION
**Pull request by rorour**
This branch is based on 3654-moderate-jupyterhub-user-creations. Please review and merge #4211 first.

### CODE REVIEW

**Comment by e-carlin** [comment]
Blocked waiting on flash center to allow our non-tls secured dev endpoint to get a response. Then need to finish https://github.com/radiasoft/sirepo/pull/4224/files#diff-6872e86bbaf9783b9521a2d7bc8efbc06219ec4d6e34d64031682626e3a13ac7R24

**Comment by e-carlin** [comment]
@robnagler this is working. I'd like another review because of the security impact. 

One decision we made along the way was to not refresh tokens. Instead just have users log in again once expired. For flash the tokens expire after one hour so this will lead to an annoying ui (need to log in to flash center every hour). Should we implement using refresh tokens or go with just re-login?

**Comment by robnagler** [comment]
Yes implement refresh tokens. It's unfortunately that they didn't set the default to something higher. First find out how many refreshes you get. Also, a refresh only works, I believe, if the user is active. Therefore, even if we implement refreshes, if they close their window and open it again, it might require a relogin

**Comment by e-carlin** [comment]
Entering invalid credentials and clicking the "deny access" button on the flash login page leads to what looks like an approved authentication response (I would expect the ui to say username/password incorrect or at the very least a response with a payload saying the request was not approved). I've reached out to the flash folks and I'm waiting on a response. This blocks the release.

**Comment by e-carlin** [comment]
@robnagler I think this is ready to merge. Please give it another review.

File: sirepo/auth/__init__.py
Comment (robnagler) [inline sirepo/auth/__init__.py:145]: This is the only thing I'm concerned about. LMK what you think.

File: sirepo/auth_role_moderation.py
Comment (robnagler) [inline sirepo/auth_role_moderation.py:144]: possibly `check_user()`?

File: sirepo/auth/__init__.py
Comment (robnagler) [inline sirepo/auth/__init__.py:138]: possibly `check_sim_type()`

**Review by robnagler — CHANGES_REQUESTED** [review]
It was easier to edit. I think the code is right. I removed some proprietary information.

File: sirepo/auth/__init__.py
Comment (e-carlin) [reply sirepo/auth/__init__.py]: That doesn't workin in the `api_checkAuthJupyterhub` case. That API is `allow_visitor` which means that this check is always True and we return when we should really be prompting to moderate. 

Thoughts ccaa8f691 ? I don't love adding another api_perm and I struggled to find a name for it...

File: sirepo/auth_role_moderation.py
Comment (e-carlin) [reply sirepo/auth_role_moderation.py]: 4980a3bd5 I think it is fine. I prefixed it with raise_ since the other calls in that method are also prefixed with raise_

File: sirepo/sim_api/jupyterhublogin.py
Comment (robnagler) [inline sirepo/sim_api/jupyterhublogin.py:81]: `self.parse_params(req_data=PKDict(simulationType='jupyterhublogin'))`

File: sirepo/sim_api/jupyterhublogin.py
Comment (robnagler) [inline sirepo/sim_api/jupyterhublogin.py:45]: I think this is a better method to use:

`self.parse_params(type='jupyterhublogin')`

This is part of our problem. I think if this were written with a, say, sim_type parameter like other calls e.g. downloadFile, then you wouldn't need to hardwire the type, which probably should come from a constant in the module, just for cleanliness.

File: sirepo/sim_api/jupyterhublogin.py
Comment (robnagler) [inline sirepo/sim_api/jupyterhublogin.py:37]: What I think is going on is that this function is doing too much. It should be `@sirepo.api_perm.require_user` which will throw an exception, and then the code in sirepo.jupyterhub would do the parsing of the response. It's all inside the sirepo repo. Then you wouldn't have to catch anything in this module. All the exceptions would cause a Response to go back to the sirepo.jupyterhub.authenticate. The fall through case (creating the juptyerhub user) would also do the right thing.

File: sirepo/auth_role_moderation.py
Comment (robnagler) [inline sirepo/auth_role_moderation.py:148]: Use constant

File: sirepo/feature_config.py
Comment (robnagler) [inline sirepo/feature_config.py:104]: two blank lines

File: sirepo/feature_config.py
Comment (robnagler) [inline sirepo/feature_config.py:137]: frozenset()

File: sirepo/feature_config.py
Comment (robnagler) [inline sirepo/feature_config.py:142]: frozenset()

File: sirepo/sim_api/jupyterhublogin.py
Comment (e-carlin) [reply sirepo/sim_api/jupyterhublogin.py:37]: How will sirepo.jupyterhub build the urls? We have access to self.sirepo_uri but all the methods like uri_for_api and local_route access the flask object directly which we can't do because we are outside the flask context.

Comment (robnagler) [reply sirepo/sim_api/jupyterhublogin.py:37]: The schema is available, but not quite. It's what's necessary to map URIs from APIs. For now, these mappings could be hardwired in sirepo.jupyterhub, since we need to get this PR out. 

I don't like adding the new API permission. Another approach is to use self.call_api and have a "hidden" api that api_checkAuthJupyterhub calls and wraps in a proper response. One things we could do to fix the SRException mess is to include a header in the response (X-Sirepo-SRException) that contains a serialized version of the exception so the HTML doesn't need to be parsed (which isn't terrible, and could be better encapsulated).

Anyway, I prefer the first approach and to get to work pulling apart the request dependent (external) from the static stuff (schema and args).

Comment (e-carlin) [reply sirepo/sim_api/jupyterhublogin.py]: 890681e6a

File: sirepo/auth_role_moderation.py
Comment (e-carlin) [reply sirepo/auth_role_moderation.py]: 890681e6a

File: sirepo/feature_config.py
Comment (e-carlin) [reply sirepo/feature_config.py]: b824893ad

File: sirepo/feature_config.py
Comment (e-carlin) [reply sirepo/feature_config.py]: b824893ad

File: sirepo/feature_config.py
Comment (e-carlin) [reply sirepo/feature_config.py:106]: b824893ad

File: sirepo/sim_api/jupyterhublogin.py
Comment (e-carlin) [reply sirepo/sim_api/jupyterhublogin.py]: Thoughts on adding sirepo_uri? Then we can re-use the code without adding another place where we handle creation of uris

File: sirepo/sim_api/jupyterhublogin.py
Comment (e-carlin) [reply sirepo/sim_api/jupyterhublogin.py]: 890681e6a4c0784a71559f1a88c2d2665ec78e19

File: sirepo/sim_api/jupyterhublogin.py
Comment (e-carlin) [reply sirepo/sim_api/jupyterhublogin.py]: 890681e6a4c0784a71559f1a88c2d2665ec78e19

File: sirepo/sim_api/jupyterhublogin.py
Comment (e-carlin) [reply sirepo/sim_api/jupyterhublogin.py]: 890681e6a4c0784a71559f1a88c2d2665ec78e19

File: sirepo/jupyterhub.py
Comment (robnagler) [inline sirepo/jupyterhub.py:94]: I hadn't seen that before. Useful for our code.

File: sirepo/jupyterhub.py
Comment (robnagler) [inline sirepo/jupyterhub.py:60]: Does there need to be a sirepo_uri? Doesn't the redirect just work on `/` without http://....?

**Review by robnagler — COMMENTED** [review]
See question about sirepo_uri.

**Review by robnagler — APPROVED** [review]

**Comment by robnagler** [comment]
@e-carlin I'll let you merge

**Pull request merged**
